rspotify-model/src/idtypes.rs

```diff
@@ -149,99 +114,57 @@ pub enum IdError {
     InvalidId,
 }
 
-impl<T: IdType> std::fmt::Display for &Id<T> {
-    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
-        write!(f, "spotify:{}:{}", T::TYPE, &self.id)
-    }
-}
-
-impl<T> AsRef<str> for &Id<T> {
-    fn as_ref(&self) -> &str {
-        &self.id
-    }
-}
-
-impl<T> Borrow<str> for &Id<T> {
-    fn borrow(&self) -> &str {
-        &self.id
-    }
-}
-
-impl<T: IdType> std::str::FromStr for IdBuf<T> {
-    type Err = IdError;
-
-    fn from_str(s: &str) -> Result<Self, Self::Err> {
-        Id::from_id_or_uri(s).map(|id| id.to_owned())
-    }
-}
-
-impl<T: IdType> Id<T> {
-    /// Owned version of the id [`IdBuf`](crate::idtypes::IdBuf).
-    pub fn to_owned(&self) -> IdBuf<T> {
-        IdBuf {
-            _type: PhantomData,
-            id: (&self.id).to_owned(),
-        }
-    }
-
-    /// Spotify object type
-    pub fn _type(&self) -> Type {
-        T::TYPE
-    }
-
-    /// Spotify object id (guaranteed to be a string of alphanumeric characters)
-    pub fn id(&self) -> &str {
-        &self.id
-    }
+/// The main interface for an ID.
+///
+/// # Implementation note
+///
+/// Note that for IDs to be useful their trait must be object-safe. Otherwise,
+/// it wouldn't be possible to use `Vec<Box<dyn Id>>` to take multiple kinds of
+/// IDs or just `Box<dyn Id>` in case the type wasn't known at compile time.
+/// This is why this trait includes both [`Self::_type`] and
+/// [`Self::_type_static`], and why all of the static methods use `where Self:
+/// Sized`.
+///
+/// Unfortunately, since `where Self: Sized` has to be enforced, IDs cannot be
+/// simply a `str` internally because these aren't sized. Thus, IDs will have the
+/// slight overhead of having to use an owned type like `String`.
+pub trait Id: Send + Sync {
+    /// Spotify object Id (guaranteed to be a string of alphanumeric characters)
+    fn id(&self) -> &str;
+
+    /// The type of the ID. The difference with [`Self::_type_static`] is that
+    /// this method can be used so that `Id` is an object-safe trait.
+    fn _type(&self) -> Type;
+
+    /// The type of the ID, which can be used without initializing it
+    fn _type_static() -> Type
+    where
+        Self: Sized;
+
+    /// Initialize the Id without checking its validity.
+    ///
+    /// # Safety
+    ///
+    /// The string passed to this method must be made out of alphanumeric
+    /// numbers only; otherwise undefined behaviour may occur.
+    unsafe fn from_id_unchecked(id: &str) -> Self
```

Comment:
Sorry for not getting your point, what are the use cases of `_type`, `_type_static` and  `from_id_unchecked`, are they necessary?

Reply:
So, the thing is that `Id` is now a trait. And just like how the Rspotify clients work, its structs need to provide getters and similars so that the implementations in `Id` work ""generically"".

* `from_id_unchecked` serves as a constructor. Since the user shouldn't initialize the ID directly (they should use `from_id` and similars), it's marked as unsafe. But otherwise there would be no way to construct the Id within the trait (how else would `from_id` be implemented, for instance?). Note that we need object safety, so it uses `where Self: Sized`
* `_type` serves as a getter for the `Type`, and so does `_type_static`. The main difference is that `_type` is a method, and `_type_static` is a static function (it doesn't take `self`). This is because there are two different use-cases to get the type: in constructors, which are static, and in regular methods:
  
  * In the `from_id` constructor if we want to get the type of the ID, we need a static function that provides it. We can't use a method that takes `&self` because the type is not constructed yet. So we just call `Self::_type_static`.
  * In `uri` if we want to get the type of the ID, we need either a static function or a method. Any of these would actually work, because the type is already constructed. The problem here is that we want to be able to use this method with a `dyn Id`. Which requires that `uri` is in `Id`'s vtable. So we can't use a static function in this case, `uri` must be a method. And in order for `uri` to be a method, it can't make calls to static functions. So `Self::_type_static` won't work, we need `self._type`.

  You can see yourself how `_type` is necessary by trying to use `_type_static` in `uri` instead. You'll get a compiler error that explains it:


  ```rust
  fn uri(&self) -> String {
      format!("spotify:{}:{}", Self::_type_static(), self.id())
  }
  ```

  results in:

  ```
  error[E0277]: the size for values of type `Self` cannot be known at compilation time
     --> rspotify-model/src/idtypes.rs:159:34
      |
  159 |         format!("spotify:{}:{}", Self::_type_static(), self.id())
      |                                  ^^^^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
      |
  note: required by `idtypes::Id::_type_static`
     --> rspotify-model/src/idtypes.rs:140:5
      |
  140 | /     fn _type_static() -> Type
  141 | |     where
  142 | |         Self: Sized;
      | |____________________^
  help: consider further restricting `Self`
      |
  158 |     fn uri(&self) -> String where Self: Sized {
      |                             ^^^^^^^^^^^^^^^^^
  
  For more information about this error, try `rustc --explain E0277`.
  error: could not compile `rspotify-model` due to previous error
  ```

Reply:
I think I (almost) understand your design, these `from_id`, `from_uri` and `from_id_or_uri` functions both are static constructor, all `Id` types can call these constructors without warring about `vtable`(of course you could make these functions as virtual function, and all `Id` types implement them, but it's clumsy).

As for `type` and `id` methods, it's necessary to make them as virtual function, while they vary between different `Id`

Reply:
> I think I (almost) understand your design, these from_id, from_uri and from_id_or_uri functions both are static constructor, all Id types can call these constructors without warring about vtable(of course you could make these functions as virtual function, and all Id types implement them, but it's clumsy).

Exactly. There's no point on making these constructors virtual functions in the vtable because you couldn't use them. `dyn Id` requires an already initialized `Id`, so a virtual constructor makes no sense AFAIK. What we want is methods like `uri` to be virtual so that we can use them in the endpoints when e.g. taking `Vec<dyn Id>`.

> As for type and id methods, it's necessary to make them as virtual function, while they vary between different Id

Yup.